core/lib/utils/src/env.rs

```diff
@@ -0,0 +1,44 @@
+use std::path::PathBuf;
+
+use once_cell::sync::OnceCell;
+
+static WORKSPACE: OnceCell<PathBuf> = OnceCell::new();
+
+/// Find the location of the current workspace, if this code works in workspace
+/// then it will return the correct folder if, it's binary e.g. in docker container
+/// you have to use fallback to another directory
+/// The code has been inspired by `insta`
+/// `https://github.com/mitsuhiko/insta/blob/master/insta/src/env.rs`
+pub fn locate_workspace() -> Option<PathBuf> {
+    let workspace = WORKSPACE.get().cloned();
+    if let Some(workspace) = workspace {
+        return Some(workspace);
+    };
+
+    let output = std::process::Command::new(
+        std::env::var("CARGO")
+            .ok()
+            .unwrap_or_else(|| "cargo".to_string()),
+    )
+    .arg("locate-project")
+    .arg("--workspace")
+    .output()
+    .map_err(|err| {
+        tracing::trace!("Can't find workspace location: {}", err);
+        err
+    })
+    .ok()?;
+
+    let root = serde_json::from_slice::<serde_json::Value>(&output.stdout)
+        .ok()?
+        .get("root")
+        .cloned()?;
+
+    let serde_json::Value::String(root) = root else {
+        return None;
+    };
+
+    let val = PathBuf::from(root).parent()?.to_path_buf();
+    WORKSPACE.set(val.clone()).unwrap();
```

Comment:
Does [`OnceCell::get_or_try_init()`](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html#method.get_or_try_init) not work here for some reason? (Shortcutting e.g., if the output JSON can't be parsed or doesn't contain `root` isn't really such a reason; it's an error that should be logged.) `set().unwrap()` can panic if called from multiple threads in parallel.